exonum/src/keys/mod.rs

```diff
@@ -0,0 +1,270 @@
+    }
+
+    /// Service public key.
+    pub fn service_pk(&self) -> PublicKey {
+        self.service.public_key()
+    }
+
+    /// Service secret key.
+    pub fn service_sk(&self) -> &SecretKey {
+        &self.service.secret_key()
+    }
+}
+
+fn save_master_key<P: AsRef<Path>, W: AsRef<[u8]>>(
+    path: P,
+    pass_phrase: W,
+    key: &secret_tree::Seed,
+) -> Result<(), Error> {
+    let encrypted_key = EncryptedMasterKey::encrypt(key, pass_phrase)?;
+    let file_content =
+        toml::to_string_pretty(&encrypted_key).map_err(|e| Error::new(ErrorKind::Other, e))?;
+    let mut open_options = OpenOptions::new();
+    open_options.create(true).write(true);
+    // By agreement we use the same permissions as for SSH private keys.
+    #[cfg(unix)]
+    open_options.mode(0o_600);
+    let mut file = open_options.open(path.as_ref())?;
+    file.write_all(file_content.as_bytes())?;
+
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize)]
+struct EncryptedMasterKey {
+    key: ErasedPwBox,
+}
+
+impl EncryptedMasterKey {
+    fn encrypt(
+        key: &secret_tree::Seed,
+        pass_phrase: impl AsRef<[u8]>,
+    ) -> Result<EncryptedMasterKey, Error> {
+        let mut rng = thread_rng();
+        let mut eraser = Eraser::new();
+        eraser.add_suite::<Sodium>();
+        let pwbox = Sodium::build_box(&mut rng)
+            .seal(pass_phrase, key)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't create a pw box"))?;
+        let encrypted_key = eraser
+            .erase(&pwbox)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't convert a pw box"))?;
+
+        Ok(EncryptedMasterKey { key: encrypted_key })
+    }
+
+    fn decrypt(self, pass_phrase: impl AsRef<[u8]>) -> Result<secret_tree::Seed, Error> {
+        let mut eraser = Eraser::new();
+        eraser.add_suite::<Sodium>();
+        let restored = eraser
+            .restore(&self.key)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't restore a secret key"))?;
+        assert_eq!(restored.len(), SEED_LENGTH);
+        let seed_bytes = restored
+            .open(pass_phrase)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't open an encrypted key"))?;
+
+        let mut seed: [u8; 32] = [0; 32];
+        seed.copy_from_slice(&seed_bytes[..]);
+        Ok(seed)
+    }
+}
+
+/// Creates a TOML file that contains encrypted master and returns `Keys` derived from it.
+pub fn generate_keys<P: AsRef<Path>>(path: P, passphrase: &[u8]) -> Result<Keys, failure::Error> {
+    let tree = SecretTree::new(&mut thread_rng());
+    save_master_key(path, passphrase, tree.seed())?;
+    generate_keys_from_master_password(tree)
+        .ok_or_else(|| format_err!("Error deriving keys from master key."))
+}
+
+fn generate_keys_from_master_password(tree: SecretTree) -> Option<Keys> {
+    let mut buffer = [0_u8; 32];
+
+    tree.child(Name::new("consensus")).fill(&mut buffer);
+    let seed = Seed::from_slice(&buffer)?;
```

Comment:
This isn't directly related to this PR, but I'm not quite fond that unlike lower-level `sodiumoxide` types, `Seed`s in `exonum-crypto` cannot be initialized securely from the external source. I'm talking something like this:

```rust
let mut seed = Seed::new([0; SEED_LENGTH]);
tree.child(Name::new("consensus")).fill(&mut seed.0);
// We know that `Seed` is zeroed on drop, so at no point in time,
// seed bytes are contained in a non-secure container.
```

Alternatively, I could imagine `Seed` API like

```rust
impl Seed {
    pub fn generate<R: RngCore + CryptoRng>(rng: &mut R) -> Self { /* ... */ }
}
```